File: ios/KOEON/Core/PTT/PttHapticFeedbackController.swift
import Foundation
import UIKit

enum HapticType: String, Equatable, Sendable {
    case press
    case release
}

enum HapticResult: String, Equatable, Sendable {
    case played
    case unsupported
    case disabled
    case failed
    case skippedDuplicate = "skipped_duplicate"
    case notPlayed = "not_played"
}

struct HapticSnapshot: Sendable {
    var supported = false
    var enabled = false
    var inputPressed = false
    var lastType: HapticType?
    var lastAt: Date?
    var lastResult: HapticResult = .notPlayed
}

@MainActor
protocol PttHapticPerforming: AnyObject {
    var supported: Bool { get }
    var enabled: Bool { get }
    func prepare()
    func performPress() throws -> Bool
    func performRelease() throws -> Bool
}

@MainActor
final class UIKitPttHapticPerformer: PttHapticPerforming {
    private let pressGenerator = UIImpactFeedbackGenerator(style: .rigid)
    private let releaseGenerator = UIImpactFeedbackGenerator(style: .light)

    let supported = true
    let enabled = true

    func prepare() {
        pressGenerator.prepare()
        releaseGenerator.prepare()
    }

    func performPress() throws -> Bool {
        pressGenerator.impactOccurred()
        pressGenerator.prepare()
        return true
    }

    func performRelease() throws -> Bool {
        releaseGenerator.impactOccurred(intensity: 0.55)
        releaseGenerator.prepare()
        return true
    }
}

/// Input feedback only: the result never controls Floor, Cue, or microphone state.
@MainActor
final class PttHapticFeedbackController {
    private let performer: any PttHapticPerforming
    private let now: () -> Date
    private let onUpdate: (HapticSnapshot) -> Void
    private var inputPressed = false
    private(set) var snapshot: HapticSnapshot

    init(
        performer: any PttHapticPerforming,
        now: @escaping () -> Date = Date.init,
        onUpdate: @escaping (HapticSnapshot) -> Void = { _ in }
    ) {
        self.performer = performer
        self.now = now
        self.onUpdate = onUpdate
        snapshot = HapticSnapshot(supported: performer.supported, enabled: performer.enabled)
    }

    func prepare() {
        try? prepareSafely()
        emit(snapshot)
    }

    @discardableResult
    func press(eligible: Bool) -> Bool {
        guard eligible else { return false }
        guard !inputPressed else {
            record(type: .press, result: .skippedDuplicate)
            return false
        }
        inputPressed = true
        play(type: .press) { try performer.performPress() }
        return true
    }

    @discardableResult
    func release() -> Bool {
        guard inputPressed else {
            record(type: .release, result: .skippedDuplicate)
            return false
        }
        inputPressed = false
        play(type: .release) { try performer.performRelease() }
        return true
    }

    func cancel() {
        guard inputPressed else { return }
        inputPressed = false
        var next = snapshot
        next.inputPressed = false
        emit(next)
    }

    private func prepareSafely() throws {
        performer.prepare()
    }

    private func play(type: HapticType, action: () throws -> Bool) {
        let result: HapticResult
        if !performer.supported {
            result = .unsupported
        } else if !performer.enabled {
            result = .disabled
        } else {
            do {
                result = try action() ? .played : .failed
            } catch {
                result = .failed
            }
        }
        record(type: type, result: result)
    }

    private func record(type: HapticType, result: HapticResult) {
        var next = snapshot
        next.supported = performer.supported
        next.enabled = performer.enabled
        next.inputPressed = inputPressed
        next.lastType = type
        next.lastAt = now()
        next.lastResult = result
        emit(next)
    }

    private func emit(_ value: HapticSnapshot) {
        snapshot = value
        onUpdate(value)
    }
}

File: android/app/src/main/java/com/dennomuso/koeon/core/haptics/PttHapticFeedbackController.kt
package com.dennomuso.koeon.core.haptics

enum class HapticType {
    PRESS,
    RELEASE,
}

enum class HapticResult(val diagnosticValue: String) {
    PLAYED("played"),
    UNSUPPORTED("unsupported"),
    DISABLED("disabled"),
    FAILED("failed"),
    SKIPPED_DUPLICATE("skipped_duplicate"),
    NOT_PLAYED("not_played"),
}

data class HapticSnapshot(
    val supported: Boolean = false,
    val enabled: Boolean = false,
    val inputPressed: Boolean = false,
    val lastType: HapticType? = null,
    val lastAtEpochMs: Long? = null,
    val lastResult: HapticResult = HapticResult.NOT_PLAYED,
)

interface HapticPerformer {
    val supported: Boolean
    val enabled: Boolean
    fun prepare()
    fun performPress(): Boolean
    fun performRelease(): Boolean
}

/** Input feedback is deliberately independent of Floor, Cue, and microphone state. */
class PttHapticFeedbackController(
    private val performer: HapticPerformer,
    private val nowEpochMs: () -> Long = System::currentTimeMillis,
    private val onSnapshot: (HapticSnapshot) -> Unit = {},
) {
    private var inputPressed = false
    private var snapshot = HapticSnapshot(
        supported = performer.supported,
        enabled = performer.enabled,
    )

    fun current(): HapticSnapshot = snapshot

    fun prepare() {
        runCatching { performer.prepare() }
        emit(snapshot.copy(supported = performer.supported, enabled = performer.enabled))
    }

    fun press(eligible: Boolean): Boolean {
        if (!eligible) return false
        if (inputPressed) {
            record(HapticType.PRESS, HapticResult.SKIPPED_DUPLICATE)
            return false
        }
        inputPressed = true
        play(HapticType.PRESS) { performer.performPress() }
        return true
    }

    fun release(): Boolean {
        if (!inputPressed) {
            record(HapticType.RELEASE, HapticResult.SKIPPED_DUPLICATE)
            return false
        }
        inputPressed = false
        play(HapticType.RELEASE) { performer.performRelease() }
        return true
    }

    fun cancel() {
        if (!inputPressed) return
        inputPressed = false
        emit(snapshot.copy(inputPressed = false))
    }

    private fun play(type: HapticType, action: () -> Boolean) {
        val result = when {
            !performer.supported -> HapticResult.UNSUPPORTED
            !performer.enabled -> HapticResult.DISABLED
            else -> runCatching { action() }
                .fold(
                    onSuccess = { if (it) HapticResult.PLAYED else HapticResult.FAILED },
                    onFailure = { HapticResult.FAILED },
                )
        }
        record(type, result)
    }

    private fun record(type: HapticType, result: HapticResult) {
        emit(
            snapshot.copy(
                supported = performer.supported,
                enabled = performer.enabled,
                inputPressed = inputPressed,
                lastType = type,
                lastAtEpochMs = nowEpochMs(),
                lastResult = result,
            ),
        )
    }

    private fun emit(value: HapticSnapshot) {
        snapshot = value
        onSnapshot(value)
    }
}
